2019 04 18 holonix (#1292)
holochain/holochain-rust: holonix/README.md

# Holonix

> NixOS and `nix-shell` in other OS environments implements the HC ops tooling.

Reviewer: ```suggestion
Comprehensive Holochain Core ops tooling (build/test/release) for use in nix compatible environments (NixOS and anywhere `nix-shell` runs).
```

## Scope

Where `nix-shell` is supported Holonix is the standard approach for:

- Installing dependencies
- Running tests
- Building wasm
- Building binaries
- Implementing CI scripts
- Conductor management
- Installing situational tooling (e.g. `cargo-edit` for versioning)
- Release management, automation and deploying official binaries
- BAU automation and reporting tasks**
- Downloading prebuilt official binaries from github**
- Managing and updating binaries across official releases**

> ** Coming Soon!

Reviewer: This seems a bit out of place, what's coming soon?

Author: @zippy the two things with `**` next to them!

### Supported environments

NixOS is a dedicated linux distro and so is standalone by its nature.

`nix-shell` is a cli tool supported across Mac OS and Linux distros.

https://nixos.org/nix/download.html

There is theoretical support for `nix-shell` in Windows Subsystem Linux (WSL).
WSL support is currently broken upstream https://github.com/NixOS/nix/issues/1203

**Use `nix-shell` if you are on anything other than Windows!**

### Why nix?

Nix approach offers unique benefits:

- Dependencies are injected into a single shell session only
  - Minimal modifications to the host environment
  - No need to maintain/rerun/troubleshoot installation scripts
  - Further isolation from host environment can be achieved with `nix-shell --pure`
- Dependencies are hashed
  - "Dependency hell" is avoided

>   - Nice parallels with the hashed zomes model

Reviewer: ```suggestion
  - Nice parallels with Holochain's hashed zomes model
```

  - Security + reliability benefits
- Dependencies can be garbage collected with the `nix-collect-garbage` command
- Single "package manager" across most operating systems
- Ability to ship utility scripts in the `shell.nix` file
- Access to the nix functional programming language for dependencies/script management
  - Allows for a structured approach to coding the ops infrastructure

>   - Modern programming paradigm such as immutability, scopes, data structures, etc.

Reviewer: ```suggestion
  - Makes modern programming paradigm such as immutability, scopes, data structures, etc. in the ops tooling as opposed limitations and lack of expressivity of Bash/Makefile/etc.
```

- NixOS runs on HoloPorts so `nix-shell` provides similar behaviour/environment
  - Everything is pinned to the same nix channel and commit

And other benefits:

- Active and helpful community of contributors
  - IRC at #nixos
  - wiki at https://nixos.wiki/
- Ability to bundle scripts into convenient binaries for daily tasks
- Nice CLI interface with many flexible, well documented configuration options
- Available in shebang form as `#! /usr/bin/env nix-shell`
  - http://iam.travishartwell.net/2015/06/17/nix-shell-shebang/

## Installation

Follow the instructions to install `nix-shell` from:

https://nixos.org/nix/download.html

Once installed run `nix-shell` from the `holochain-rust` repository root.
The `nix-shell` command will detect the `shell.nix` file and build everything.

The first run will be slow as nix downloads, builds and caches dependencies.
Everything is downloaded to `/nix/store/...` and re-used on subsequent runs.

> ## Contributing

Reviewer: I think this contributing section could also use a bit of instruction on the `dist` folders which it looks like are used to hold derivations/scripts for build artifacts for release.  Correct?

Author: this isn't "contributing to holochain core" it is "contributing to holonix"

The structure of our nix derivations is designed to be as modular as possible.

The folder structure is a basic heirarchy, something like:

```
holonix
 |_github
 |  |_config.nix
 |_nixpkgs
 | |_nixpkgs.nix
 |_node
 | |_src
 | | |_flux.nix
 | | |_...
 | |_build.nix
 |_...
shell.nix
```

The `shell.nix` file is used by `nix-shell` automatically by default.
This consumes `holonix/**` and does not provide any new derivations.

> End-users should not need to interact with `holonix/**` outside `shell.nix`.

Reviewer: This is a weird sentence here because it's in the context of "Contributing"  Maybe it should be move to the top?  Also who is the "end-user"  Holochain DNA devs?

Author: contributing to holonix, end users are users of holonix

> There are a few basic convensions to follow:

Reviewer 2: typo?
```suggestion
There are a few basic conventions to follow:
```

- Nest folders according to theme/tech/specificity
  - e.g. conductor management for conductor `x` sits under `conductor/x/**`
- All configuration strings and other primitives sit in a local `config.nix`
- Structure configuration as nested `foo.bar` rather than `foo-bar`
  - e.g. `holonix/release/config.nix` has a few good examples of this

> - All used and generated build inputs sit in a local `build.nix`

Reviewer: I think "build" here refers to building up the derivations, not building holochain artifacts correct?  If so I'd suggest the following change:

```suggestion
- All used and generated inputs to build the nix derivations sit in a local `build.nix`
```

But I might have that wrong.

Author: yeah i mean `buildInputs` for `mkDerivation`

  - `build.nix` files should "bubble up" to the root one level at a time
    - e.g. `build.nix` imports `conductor/build.nix` imports `conductor/node/build.nix`
  - root `build.nix` and `src` should only aggregate deeper derivations
- Scripts for binaries sit in named `foo.nix` files under `thing/src/foo.nix`

>   - There is standard boilerplate for this, see an existing file for examples

Reviewer: Can you point to an actual example file here?  "thing/src/foo.nix" is so general.  What's the "thing"?  the binary?  Or is "foo" the binary?

Author: oooh, see i'm talking about `writeShellScriptBin` (see below) where nix treats the shell script as a binary

this doesn't refer to binaries like `holochain` and `hc`, it just means that if you had a "binary" command called `hc-thing-foo` you'd put it on the file path `thing/src/foo.nix`

  - Use `pkgs.writeShellScriptBin` by default

>   - CLI commands are named following the path sans `src`

Reviewer: ```suggestion
  - derived nix CLI commands are named following the path sans `src`
```

(so as to distinguish from holochain cli commands)

    - e.g. `holonix/foo/bar/src/baz.nix` becomes `hc-foo-bar-baz`

> - Make liberal use of `let .. in ..` scoping constructs in `.nix` files

Reviewer: maybe an explanatory clause about this injunction:

- So as to X, please make liberal use of `let .. in ..` scoping constructs in `.nix` files

Author: oh uh, i'm not sure how to explain it, maybe we can discuss?
